src/lib.rs

```diff
@@ -97,16 +97,118 @@
 
 mod collector;
 
+use std::{mem, sync::Arc};
+
 use metrics::{describe_gauge, gauge, Unit};
 
+/// Metrics names
+#[derive(Debug, PartialEq, Eq)]
+struct Metrics {
+    prefixed: bool,
+    cpu_seconds_total: &'static str,
+    open_fds: &'static str,
+    max_fds: &'static str,
+    virtual_memory_bytes: &'static str,
+    virtual_memory_max_bytes: &'static str,
+    resident_memory_bytes: &'static str,
+    start_time_seconds: &'static str,
+    threads: &'static str,
```

Comment:
If we've used `String`s for those fields, we would get this error:

```rust
error[E0521]: borrowed data escapes outside of method
   --> src/lib.rs:219:23
    |
218 |     pub fn describe(&self) {
    |                     -----
    |                     |
    |                     `self` is a reference that is only valid in the method body
    |                     let's call the lifetime of this reference `'1`
219 |         let metrics = self.metrics.as_ref();
    |                       ^^^^^^^^^^^^^^^^^^^^^
    |                       |
    |                       `self` escapes the method body here
    |                       argument requires that `'1` must outlive `'static`

error[E0521]: borrowed data escapes outside of method
   --> src/lib.rs:288:13
    |
284 |     pub fn collect(&self) {
    |                    -----
    |                    |
    |                    `self` is a reference that is only valid in the method body
    |                    let's call the lifetime of this reference `'1`
...
288 |             gauge!(metrics.cpu_seconds_total.as_str()).set(v);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |             |
    |             `self` escapes the method body here
    |             argument requires that `'1` must outlive `'static`
    |
    = note: this error originates in the macro `$crate::key_var` which comes from the expansion of the macro `gauge` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Reply:
This PR provides an implementation of the alternative to `leak`ing the strings, that additionally frees them in the `Drop`.

Reply:
Obsolete https://github.com/lambdalisue/rs-metrics-process/pull/35#issuecomment-1870123042